velox/common/caching/SsdCache.h

```diff
@@ -0,0 +1,355 @@
+class SsdPin {
+ public:
+  SsdPin() : file_(nullptr) {}
+
+  // Constructs a pin referencing 'run' in 'file'. The region must be
+  // pinned before constructing the pin.
+  SsdPin(SsdFile& file, SsdRun run);
+
+  SsdPin(const SsdPin& other) = delete;
+
+  SsdPin(SsdPin&& other) {
+    run_ = other.run_;
+    file_ = other.file_;
+    other.file_ = nullptr;
+  }
+
+  ~SsdPin();
+
+  void operator=(SsdPin&&);
+
+  bool empty() const {
+    return file_ == nullptr;
+  }
+  SsdFile* file() const {
+    return file_;
+  }
+
+  SsdRun run() const {
+    return run_;
+  }
+
+ private:
+  SsdFile* file_;
+  SsdRun run_;
+};
+
+// Metrics for SSD cache. Maintained by SsdFile and aggregated by SsdCache.
+struct SsdCacheStats {
+  uint64_t entriesWritten{0};
+  uint64_t bytesWritten{0};
+  uint64_t entriesRead{0};
+  uint64_t bytesRead{0};
+  uint64_t entriesCached{0};
+  uint64_t bytesCached{0};
+};
+
+// A shard of SsdCache. Corresponds to one file on SSD.  The data
+// backed by each SsdFile is selected on a hash of the storage file
+// number of the cached data. Each file consists of an integer number
+// of 64MB regions. Each region has a pin count and an read
+// count. Cache replacement takes place region by region, preferring
+// regions with a smaller read count. Entries do not span
+// regions. Otherwise entries are consecutive byte ranges inside
+// their region.
+class SsdFile {
+ public:
+  static constexpr uint64_t kMaxSize = 1UL << 36; // 64G
+  static constexpr uint64_t kRegionSize = 1 << 26; // 64MB
+  static constexpr int32_t kNumRegions = kMaxSize / kRegionSize;
+
+  // Constructs a cache backed by filename. Discards any previous
+  // contents of filename.
+  SsdFile(
+      const std::string& filename,
+      SsdCache& cache,
+      int32_t ordinal,
+      int32_t maxRegions);
+
+  // Adds entries of  'pins'  to this file. 'pins' must be in read mode and
+  // those pins that are successfully added to SSD are marked as being on SSD.
+  // The file of the entries must be a file that is backed by 'this'.
+  void store(std::vector<CachePin>& pins);
+
+  // Finds an entry for 'key'. If no entry is found, the returned pin is empty.
+  SsdPin find(RawFileCacheKey key);
+
+  // Copies the data at 'run' into 'entry'. Checks that the entry
+  // and run sizes match. The quantization of SSD cache matches that
+  // of the memory cache.
+  void load(SsdRun run, AsyncDataCacheEntry& entry);
+
+  // Reads the backing file as per ReadFile::preadv(). 'numEntries' is
+  // used only for updating statistics. Allows implementing coalesced
+  // read of nearby entries.
+  void read(
+      uint64_t offset,
+      const std::vector<folly::Range<char*>> buffers,
+      int32_t numEntries);
+
+  // Increments the pin count of the region of 'offset'.
+  void pinRegion(uint64_t offset);
+
+  // Increments the pin count of the region of 'offset'. Caller must hold
+  // 'mutex_'.
+  void pinRegionLocked(uint64_t offset) {
```

Comment:
Should this be a private method? I don't see mutex_ exposed to the users of this class.